### ISSUE TITLE
Feature Request: structured logging format

### ISSUE THREAD

**Issue by renchap** [mperham/sidekiq#4046]
This is a best practice on production platforms to send logs to a central collector, parse them and store them as a structured document. This allows to create dashboards and alerts based on logs.

Current Sidekiq logs are plain text and need a custom parser to be processed. It would be awesome to be able to (optionally) enable a JSON (or similar, ideally modular) format. Log messages could include a type (start job / stop job), Job name, Job ID, duration (for a stop event) or even job parameters.

For Rails, the `lograge` gem provides this and is used in most production deployments: https://github.com/roidrage/lograge

Current implementations:
- https://github.com/Springest/Sidekiq-Logging-JSON
- Gitlab: https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/18121

**Comment by olivierlacan** [comment]
Although it's a bit more convoluted, the SemanticLogger gem can also convert default Ruby/Rails logs into JSON output which is what we use at Pluralsight to stream JSON to Kibana: https://github.com/rocketjob/semantic_logger

I think the logging could be improved by providing essentials like: 
- perform duration
- completed or failed (exception)

In the `sidekiq.rb` initializer this is how I currently override Sidekiq's JobLogger:

```ruby
if Rails.logger.is_a?(SemanticLogger::Logger)
  module Sidekiq
    class JobLogger
      def call(_item, _queue)
        start = Time.current

        logger.info('start'.freeze)

        yield

        duration = elapsed(start)

        logger.info(message: "done: #{duration} sec", payload: {
          duration: duration * 1000
        })
      rescue StandardError
        duration = elapsed(start)

        logger.info(message: "fail: #{duration} sec", payload: {
          duration: duration * 1000
        })

        raise
      end
    end
  end
end
```

**Comment by Tensho** [comment]
I'd definitely like to work on this for sure, guys. But I'd like to understand the problem we would like to solve. There are 2 aspects we have to consider:
- `Sidekiq.logger` (`Sidekiq::Logging.logger`) is a **general purpose logger**, exists to suport  multi-threaded system logging
- `Sidekiq::JobLogger` is a specific logger, repsonsible only for the job lifecycle logging (elapsed time), prime consumer of `Sidekiq.logger`.

As I understand from the issue request it would be nice to **get structured logging formatter for `Sidekiq::JobLogger` out of the box (without any extra dependencies)**. We don't care much about banner and other configuration info in JSON form here.

I've inspected both [Sidekiq-Logging-JSON](https://github.com/Springest/Sidekiq-Logging-JSON) and [semantic_logger](https://github.com/rocketjob/semantic_logger). The first is very simple (ruby hash) and I guess it would be nice to start with the something similar in Sidekiq core. The second is a swiss-army knife 🇨🇭 🔪   which [supports Sidekiq](https://github.com/rocketjob/semantic_logger/blob/master/docs/api.md#545) out of the box.

@mperham Please provide your insight on the subject. As far as Sidekiq Logging API allows to provide a custom logger and there is a bunch of mature frameworks around, does it worth efforts? I see the main advantage in service-specific (logstash, graylog) formatters for people outside, but not in a raw one.

@olivierlacan ATM job logger provides [elapsed time and state (`done`/`fail`)](https://github.com/mperham/sidekiq/blob/master/lib/sidekiq/job_logger.rb#L9).

**Comment by jfelchner** [comment]
You all can also check out my gem: https://github.com/thekompanee/juan_pelota

**Comment by mperham** [comment]
Sidekiq::Logging is used by a lot of user code.  I would prefer to keep compatibility unless you have a good reason to break apps.

**Comment by renchap** [comment]
> I'd definitely like to work on this for sure, guys. But I'd like to understand the problem we would like to solve. There are 2 aspects we have to consider:
> 
> * `Sidekiq.logger` (`Sidekiq::Logging.logger`) is a **general purpose logger**, exists to suport  multi-threaded system logging
> * `Sidekiq::JobLogger` is a specific logger, repsonsible only for the job lifecycle logging (elapsed time), prime consumer of `Sidekiq.logger`.
> 
> As I understand from the issue request it would be nice to **get structured logging formatter for `Sidekiq::JobLogger` out of the box (without any extra dependencies)**. We don't care much about banner and other configuration info in JSON form here.

If Sidekiq logs everything on stderr/stdout, I would like to have all lines to be JSON if Sidekiq is configured to be JSON. Otherwise you need again to have a custom log processor configuration to separate the JSON lines from the others.

I agree that the banner and some other lines are only useful when the output is plain text, but the few other messages should be formatted in JSON. For example `INFO: Booting Sidekiq 5.2.3 with redis options {:id=>"Sidekiq-server-PID-5495", :url=>nil}` might be useful` while I dont see any use for `INFO: Upgrade to Sidekiq Pro for more features and support: http://sidekiq.org`.
Logging the startup / shutdown phases is also useful to monitor Sidekiq restarts.

**Comment by Tensho** [comment]
@mperham The main problem I see there are 2 context – generic and job logging. **Only** job log entries should contain job elapsed time and JID, but not generic. Does `config.log_format = :json` set JSON formatter globally? Should we separate formatters for the different loggers?

I agree with @renchap JSON formatter should be global and be able to output system or job related info on demand.

I wonder Sidekiq encourages people to monkey patch the class (`#`). Regarding backward compatibility, we may proxy to the new places:
```
# sidekiq/logging.rb
module Sidekiq
  class Logging
    puts "WARNING: This class proxies to Sidekiq main module and will be removed soon in favor of Sidekiq::Logger"

    def self.logger
      Sidekiq.logger
    end

    def self.logger=(logger)
      Sidekiq.logger = logger
    end

    def self.tid
      Sidekiq.tid
    end

    def self.job_hash_context(job_hash)
      JobLogger.new.job_hash_context(job_hash)
    end

    def self.with_job_hash_context(job_hash, &block)
      Sidekiq.logger.with_context(job_hash_context(job_hash), &block)
    end
  end
end
```
I may cover user scenarios with specs if you give me examples.

**Comment by mperham** [comment]
Keep it simple. There should be only one process-wide logger and the format should be set during initial configuration.  If you configure :json, every line from that point forward should be json. The log output could be as simple as `{"pid":12345,"level":"INFO", "msg":"Something informative"}` based on the context in which it is used.

**Comment by Tensho** [comment]
@mperham How could one process-wide logger get job elapsed time along with the message? Please, could you provide logger config and API for pretty and json formatters please? I see the only one way to pass arbitrary data on demand – context:
```
Sidekiq.logger.with_context(elapsed: 5.2) do
  Sidekiq.logger.info "done"
end
```
Does it make sense?

**Comment by mperham** [comment]
Yes, a thread-local variable will store the contextual data for each thread, exactly as you show via `with_context`.

**Comment by Tensho** [comment]
@mperham OK, I'll implement that. What about `Sidekiq::Logger` backward compatibility? Are you OK with my proposal or would you prefer to be conservative and leave `Sidekiq::Logging`?

**Comment by mperham** [comment]
I'm ok with breaking Sidekiq::Logging as long as this targets Sidekiq 6.0 and the upgrade notes mention the breakage.

**Comment by Tensho** [comment]
Implemented in #4050, will be released in `Sidekiq 6.0`.
